auth/sasl/authenticatee.go

```diff
@@ -0,0 +1,320 @@
+	credential mesos.Credential    // user-specified credentials
+	transport  messenger.Messenger // mesos communications transport
+}
+
+func nextPid() int {
+	pidLock.Lock()
+	defer pidLock.Unlock()
+	pid++
+	return pid
+}
+
+func (s *statusType) get() statusType {
+	return statusType(atomic.LoadInt32((*int32)(s)))
+}
+
+func (s *statusType) swap(old, new statusType) bool {
+	return old != new && atomic.CompareAndSwapInt32((*int32)(s), int32(old), int32(new))
+}
+
+func Authenticatee(ctx context.Context, pid, client upid.UPID, credential mesos.Credential) <-chan error {
+
+	c := make(chan error, 1)
+	f := func() error {
+		config := newConfig(client, credential)
+		ctx, auth := newAuthenticatee(ctx, config)
+		auth.authenticate(ctx, pid)
+
+		select {
+		case <-ctx.Done():
+			return auth.discard(ctx)
+		case <-auth.done:
+			return auth.result
+		}
+	}
+	go func() { c <- f() }()
+	return c
+}
+
+func newConfig(client upid.UPID, credential mesos.Credential) *authenticateeConfig {
+	tpid := &upid.UPID{ID: fmt.Sprintf("sasl_authenticatee(%d)", nextPid())}
+	return &authenticateeConfig{
+		client:     client,
+		credential: credential,
+		transport:  messenger.NewMesosMessenger(tpid),
+	}
+}
+
+// Terminate the authentication process upon context cancellation;
+// only to be called if/when ctx.Done() has been signalled.
+func (self *authenticateeProcess) discard(ctx context.Context) error {
+	err := ctx.Err()
+	status := statusFrom(ctx)
+	for ; status < _TERMINAL; status = (&self.status).get() {
+		if self.terminate(status, DISCARDED, err) {
+			break
+		}
+	}
+	return err
+}
+
+func newAuthenticatee(ctx context.Context, config *authenticateeConfig) (context.Context, *authenticateeProcess) {
+	initialStatus := READY
+	proc := &authenticateeProcess{
+		transport:  config.transport,
+		client:     config.client,
+		credential: config.credential,
+		status:     initialStatus,
+		done:       make(chan struct{}),
+	}
+	ctx = withStatus(ctx, initialStatus)
+	err := proc.installHandlers(ctx)
+	if err == nil {
+		err = proc.startTransport()
+	}
+	if err != nil {
+		proc.terminate(initialStatus, ERROR, err)
+	}
+	return ctx, proc
+}
+
+func (self *authenticateeProcess) startTransport() error {
+	if err := self.transport.Start(); err != nil {
+		return err
+	} else {
+		go func() {
+			// stop the authentication transport upon termination of the
+			// authenticator process
+			select {
+			case <-self.done:
+				log.V(2).Infof("stopping authenticator transport: %v", self.transport.UPID())
+				self.transport.Stop()
+			}
+		}()
+	}
+	return nil
+}
+
+// returns true when handlers are installed without error, otherwise terminates the
+// authentication process.
+func (self *authenticateeProcess) installHandlers(ctx context.Context) error {
+
+	type handlerFn func(ctx context.Context, from *upid.UPID, pbMsg proto.Message)
+
+	withContext := func(f handlerFn) messenger.MessageHandler {
+		return func(from *upid.UPID, m proto.Message) {
```

Comment:
would be nice to get rid of another TODO, check `from` against an optional `fromFilter` here, to ensure that we keep talking to the same authenticator